### PR TITLE
Record telemetry for nimbus experiments fetching

### PR DESCRIPTION
**Pull request by ionutbedregeanu**
For #6847
Add a timespan to know how much it takes to download the experiments
No tests, screenshots, accessibility required
### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on `Show All Checks`,
2. click `Details` next to `build-focus-debug` or `build-klar-debug` for changes targeting Klar,
2. click `View task in Taskcluster`,
3. click the `Artifacts` row,
4. click to download any of the apks listed here which use an appropriate name for each CPU architecture.

### CODE REVIEW

**Comment by ionutbedregeanu** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive a review from a data steward peer on your responses to these questions before shipping new data collection.**

_1) What questions will you answer with this data?_
How much it takes to fetch the nimbus experiments?

_2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?_
This is needed to analyze feature usage and possible feature improvements.

_3) What alternative methods did you consider to answer these questions? Why were they not sufficient?_
This is the only way.

_4) Can current instrumentation answer these questions?_
No. We need a new specific event for a specific app feature.

_5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Firefox/Data_Collection) found on the Mozilla wiki._

_**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**_

<table>
  <tr>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>

  <tr>
    <td>Nimbus experiments timespan</td>
    <td>Category 2</td>
    <td>https://github.com/mozilla-mobile/focus-android/issues/6847</td>
    </tr>

</table>

_6) How long will this data be collected?  Choose one of the following:_
    Expiry for capturing data is set to release version 112

_7) What populations will you measure?_
All release channels and locales.

_8) If this data collection is default on, what is the opt-out mechanism for users?_
 Users can opt out of data collection by disabling Usage and technical data from Settings -> Privacy and security -> Data choices.

_9) Please provide a general description of how you will analyze this data._
 Glean 

_10) Where do you intend to share the results of your analysis?_
Only on Glean, mobile teams, and BD have internal access.

_12) Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?_
No.

File: app/src/main/java/org/mozilla/focus/experiments/NimbusSetup.kt [hunk 1]
```diff
@@ -82,7 +84,16 @@ fun createNimbus(context: Context, url: String?): NimbusApi {
             }
 
             applyPendingExperiments()
+            NimbusExperiments.nimbusInitialFetch.start()
             fetchExperiments()
+
+            register(object : NimbusInterface.Observer {
+                override fun onExperimentsFetched() {
+                    NimbusExperiments.nimbusInitialFetch.stop()
```
Comment (Amejia481) [inline app/src/main/java/org/mozilla/focus/experiments/NimbusSetup.kt:92]: Just to clarify, based on the description of https://github.com/mozilla-mobile/focus-android/issues/6847, should we be recording the event when we are not able to fetch the experiments and going to the default experience ? :)

Comment (Amejia481) [reply app/src/main/java/org/mozilla/focus/experiments/NimbusSetup.kt:92]: Maybe there is a metric that I'm ignoring, please help me to clarify, I think we will need to also have amount of users that try to download. Having the total users that try to download and the users download successfully, I think we could figure out how many failed to download.

Comment (Amejia481) [reply app/src/main/java/org/mozilla/focus/experiments/NimbusSetup.kt:92]: Base on the documentation  for `cancel`:
> Abort a previous start call. No error is recorded if no start was called.

**Comment by ionutbedregeanu** [comment]
successful_fetch should be updated to support also that scenario when the user doesn't have internet connection

**Comment by Amejia481** [comment]
> successful_fetch should be renamed and converted to an event which should contains information about the state(failed or not), but also about the internet connection(if the device is connected to internet or not). @Amejia481 @jhugman What do you think about this?

I think a boolean it's sufficient as long as we report `true` when we are able to fetch the experiments and false otherwise.

File: app/src/main/java/org/mozilla/focus/experiments/NimbusSetup.kt (hunk 1 above)
Comment (jhugman) [reply app/src/main/java/org/mozilla/focus/experiments/NimbusSetup.kt:92]: It's worth asking **why** we're collecting this information.

We're collecting this to make a decision on how long to set the timeout. The fetch will continue even if the splash screen is dismissed (and by the looks of it here: on every startup).

I don't **think** we need to collect whether or not the initial fetch was completed in time for the onboarding.

What do you think?

**Review by jhugman — APPROVED** [review]
I'm not convinced of the utility of the `successfulFetch` event, so suggest you remove it.

Also: could you amend the data review request to make the data category "Category 1: Technical". **Edit**: I see you've done the right thing in the `metrics.yaml`!

Otherwise, this is nice, well done. Thank you.

File: app/src/main/java/org/mozilla/focus/activity/MainActivity.kt [hunk 2]
```diff
@@ -133,6 +134,7 @@ open class MainActivity : LocaleAwareAppCompatActivity() {
         content.viewTreeObserver.addOnPreDrawListener(object : ViewTreeObserver.OnPreDrawListener {
             override fun onPreDraw(): Boolean {
                 return if (System.currentTimeMillis() >= endTime) {
+                    NimbusExperiments.successfulFetch.set(false)
```
[the remark this answers is not in the document]
Comment (ionutbedregeanu) [reply app/src/main/java/org/mozilla/focus/activity/MainActivity.kt:137]: successfulFetch was added as a flag to know if the experiments were downloaded until timeout, but if the timespan is enough I can remove it

**Comment by travis79** [comment]
> 

# Data Review

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?

Yes, through the metrics.yaml file and the [Glean Dictionary](https://dictionary.telemetry.mozilla.org/).

2) Is there a control mechanism that allows the user to turn the data collection on and off?

Yes, through the "Send Usage Data" preference in the application settings.

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

N/A, collection to end in v112

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Category 1, Technical data

5) Is the data collection request for default-on or default-off?

Default-on

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?

No

7) Is the data collection covered by the existing Firefox privacy notice?

Yes

8) Does the data collection use a third-party collection tool?

No

## Result

data-review+

**Review by travis79 — COMMENTED** [review]
data-review+

**Pull request merged**